File: solutions/python/mschmid98/taschenrechner.py

class MyBuffer:
    def __init__(self, input : str):
        self._input = input
        self._pos = 0
    
    def peek(self) -> str:
        if self._pos > len(self._input):
            raise 
        return '#' if self._pos == len(self._input) else self._input[self._pos]
    
    def skip(self) -> 'MyBuffer':
        self._pos +=1
        return self
    
    def pop(self) -> str:
        r = self.peek()
        self._pos += 1
        return r
    
    @property
    def input(self) -> str:
        return self._input
    
    @property
    def pos(self) -> int:
        return self._pos

class TRSyntaxException(Exception):
    def __init__(self, input : MyBuffer, expected : str):
        self._input = input.input
        self._pos = input.pos
        self._expected = expected
        super().__init__(self._shortMessage)
    
    @property
    def _shortMessage(self) -> str:
        if self._pos ==len(self._input):
            return "Error: Missing character at end of input, expected: " + self._expected
        else:
            return "Error at column " + str(self._pos) + ": Got '" + self._input[self._pos] + "', expected: " + self._expected

    @property
    def errorString(self) -> str:
        return self._input + "\n" + " " * self._pos + "^\n" + self._shortMessage

def stateS(accum : float, input : MyBuffer) -> float:
    accum = stateE(accum, input)

    if input.peek() == '#':
        return accum
    else:
        raise TRSyntaxException(input, "<EOF>")

def stateE(accum : float, input : MyBuffer) -> float:
    accum = stateP(accum, input)
    return stateE2(accum, input)

def stateE2(accum : float, input : MyBuffer) -> float:
    if input.peek() == "+":
        accum += stateP(accum, input.skip())
        return stateE2(accum, input)
    elif input.peek() == "-":
        accum -= stateP(accum, input.skip())
        return stateE2(accum, input)
    else:
        return accum
    
def stateP(accum : float, input : MyBuffer) -> float:
    accum = stateT(accum, input)
    return stateP2(accum, input)
    
def stateP2(accum : float, input : MyBuffer) -> float:
    if input.peek() == '*':
        accum *= stateT(accum, input.skip())
        return stateP2(accum, input)
    elif input.peek() == '/':
        accum /= stateT(accum, input.skip())
        return stateP2(accum, input)
    else:
        return accum

def stateT(accum : float, input : MyBuffer) -> float:
    if input.peek() == '-':
        return -stateT(accum, input.skip())
    else:
        accum = stateF(accum, input)
        return stateT2(accum, input)

def stateT2(accum : float, input : MyBuffer) -> float:
    if input.peek() == '^':
        return accum ** stateT(accum, input.skip())
    else:
        return accum

def stateF(accum : float, input : MyBuffer) -> float:
    if input.peek() == '(':
        accum = stateE(accum, input.skip())

        if input.peek() != ')':
            raise TRSyntaxException(input, ")")
        else:
            input.skip()
        
        return accum
    else:
        return stateZ(accum, input)

def stateZ(accum : float, input : MyBuffer) -> float:
    if input.peek() != '.' and (input.peek() < '1' or input.peek() > '9'):
        raise TRSyntaxException(input, "<NON-ZERO-DIGIT>")
    
    number = 0
    frac = 0

    while input.peek() >= '0' and input.peek() <= '9':
        number = number * 10 + int(input.pop())
    
    if input.peek() == '.':
        div = 10

        if input.peek() < '0' or input.peek > '9':
            raise TRSyntaxException(input, "<DIGIT>")
        
        while input.peek() >= '0' and input.peek() <= '9':
            frac = frac + int(input.pop()) / div
            div *= 10
    
    return number + frac

def parse(s : str) -> float:
    # Geht auch sauberer via RegEx
    s = s.replace(" ", "").replace("\t", "").replace("\n", "").replace("\r", "")
    return stateS(0, MyBuffer(s))

if __name__ == '__main__':
    import sys

    if len(sys.argv) <= 1:
        print("Missing input", file=sys.stderr)
        exit(1)
    else:
        try:
            print(parse(sys.argv[1]))
        except TRSyntaxException as e:
            print(e.errorString, file=sys.stderr)
            exit(-1)

File: solutions/java/mschmid98/Taschenrechner.java
import java.nio.CharBuffer;

public final class Taschenrechner {
    private static class SyntaxException extends Exception {
        private String input;
        private int pos;
        private String expected;

        private SyntaxException(CharBuffer input, String expected) {
            super(getShortMessage(input, expected));
            this.pos = input.position();
            input.rewind();
            this.input = input.toString();
            this.expected = expected;
        }

        private static String getShortMessage(CharBuffer input, String expected) {
            int pos = input.position();
            input.rewind();
            String inString = input.toString();
            input.position(pos);

            return getShortMessage(pos, expected, inString);
        }

        private static String getShortMessage(int pos, String expected, String input) {
            StringBuilder sb = new StringBuilder();
            if (pos == input.length()) {
                sb.append("Error: Missing character at end of input, expected: ")
                    .append(expected);
            } else {
                sb.append("Error at column ").append(pos)
                    .append(": Got '").append(input.charAt(pos))
                    .append("' expected: ").append(expected);
            }
            return sb.toString();
        }

        public String getErrorString() {
            StringBuilder sb = new StringBuilder();
            sb.append(input).append('\n');

            for (int i = 0; i < pos; i++) 
                sb.append(' ');
            sb.append("^\n");
        
            sb.append(getShortMessage(pos, expected, input));
            return sb.toString();
        }
    }

    private static double stateS(double accum, CharBuffer input) throws SyntaxException {
        accum = stateE(accum, input);

        if (input.remaining() == 1 && input.get() == '#') {
            return accum;
        } else {
            throw new SyntaxException(input, "<EOF>");
        }
    }
    
    private static double stateE(double accum, CharBuffer input) throws SyntaxException {
        accum = stateP(accum, input);
        accum = stateE2(accum, input);
        return accum;
    }
    
    private static double stateE2(double accum, CharBuffer input) throws SyntaxException {
        switch (input.get()) {
            case '+':
                accum += stateP(accum, input);
                return stateE2(accum, input);
            case '-':
                accum -= stateP(accum, input);
                return stateE2(accum, input);
            default:
                input.position(input.position()-1);
                return accum;
        }
    }
    
    private static double stateP(double accum, CharBuffer input) throws SyntaxException {
        accum = stateT(accum, input);
        accum = stateP2(accum, input);
        return accum;
    }
    
    private static double stateP2(double accum, CharBuffer input) throws SyntaxException {
        switch (input.get()) {
            case '*':
                accum *= stateT(accum, input);
                return stateP2(accum, input);
            case '/':
                accum /= stateT(accum, input);
                return stateP2(accum, input);
            default:
                input.position(input.position()-1);
                return accum;
        }
    }
    
    private static double stateT(double accum, CharBuffer input) throws SyntaxException {
        switch (input.get()) {
            case '-':
                return -stateT(accum, input);
            default:
                input.position(input.position()-1);
                accum = stateF(accum, input);
                return stateT2(accum, input);
        }
    }
    
    private static double stateT2(double accum, CharBuffer input) throws SyntaxException {
        switch (input.get()) {
            case '^':
                return Math.pow(accum, stateT(accum, input));
            default:
                input.position(input.position()-1);
                return accum;
        }
    }
    
    private static double stateF(double accum, CharBuffer input) throws SyntaxException {
        switch (input.get()) {
            case '(':
                accum = stateE(accum, input);
    
                // absolute get, avoid position increase
                if (input.get(input.position()) != ')')
                    throw new SyntaxException(input, ")");
                else
                    input.get(); // Increase position
    
                return accum;
            default:
                input.position(input.position()-1);
                return stateZ(accum, input);
        }
    }
    
    private static double stateZ(double accum, CharBuffer input) throws SyntaxException {
        char current = input.get(input.position());

        if (current != '.' && (current < '1' || current > '9'))
            throw new SyntaxException(input, "<NON-ZERO-DIGIT>");
        
        double number = 0;
        double frac = 0;
    
        while (true) {
            current = input.get();
            if (current < '0' || current > '9')
                break;
            number = number * 10 + (current - '0');
        }
    
        if (current == '.') {
            double div = 10;
            current = input.get(input.position());
    
            if (current < '0' || current > '9')
                throw new SyntaxException(input, "<DIGIT>");
            
            while (true) {
                current = input.get();
                if (current < '0' || current > '9')
                    break;
                frac = frac + (current - '0') / div;
                div *= 10;
            }
        }

        input.position(input.position()-1);
        return number + frac;
    }

    public static double parse(String s) throws SyntaxException {
        CharBuffer buf = CharBuffer.wrap(s + "#");
        return stateS(0, buf);
    }
    
    public static void main(String[] args) {
        try {
            if (args.length == 0) {
                System.err.println("Missing input");
                System.exit(1);
            } else {
                // Geht auch sauberer via Regex
                String input = args[0].replace(" ", "").replace("\t", "").replace("\n", "").replace("\r", "");
                System.out.println(parse(input));
            }
        } catch (SyntaxException ex) {
            ex.printStackTrace();
            System.err.println(ex.getErrorString());
            System.exit(-1);
        }
    }
}
